File: worker/loops/l2.ts
import type Redis from "ioredis";
import type { ExploreSource } from "../agents/curator";
import { getHermes, type DiscoveryStrategy } from "../../src/lib/core/hermes";
import type { RedisLike } from "../../src/lib/core/store/types";

/**
 * L2 — Self-evolving discovery + self-grading. Candidate follow-ups are proposed
 * here from a session; the yes/no reward and versioned strategy/skill evolution
 * are handled by the Hermes module (src/lib/core/hermes), which persists living
 * markdown files to the Vault and mirrors the strategy to Redis.
 */

const suggestionsKey = (uid: string) => `suggestions:${uid}`;
const seenSourcesKey = (uid: string) => `seen_sources:${uid}`;
const seenAuthorsKey = (uid: string) => `seen_authors:${uid}`;

const STOP = new Set([
  "the", "and", "for", "with", "that", "this", "from", "what", "your", "you", "are", "how",
  "why", "new", "now", "best", "top", "vs", "use", "using", "into", "about", "over", "last",
  "days", "people", "say", "said", "want", "users", "guide", "tutorial",
]);

/** Lightweight follow-up extraction from source titles + synthesis (no extra LLM call). */
export async function proposeFollowups(
  redis: Redis,
  uid: string,
  topic: string,
  sources: ExploreSource[],
  synthesis: string
): Promise<string[]> {
  const text = [synthesis, ...sources.map((s) => `${s.title} ${s.snippet ?? ""}`)].join(" ");
  const topicWords = new Set(topic.toLowerCase().split(/\s+/));

  // Candidate phrases: Capitalized multi-word terms and notable single tokens.
  const counts = new Map<string, number>();
  const phraseRe = /\b([A-Z][a-zA-Z0-9.+#]+(?:\s+[A-Z][a-zA-Z0-9.+#]+){0,2})\b/g;
  let m: RegExpExecArray | null;
  while ((m = phraseRe.exec(text))) {
    const phrase = m[1].trim();
    const lower = phrase.toLowerCase();
    if (phrase.length < 3) continue;
    if (topicWords.has(lower)) continue;
    if (lower.split(/\s+/).every((w) => STOP.has(w))) continue;
    counts.set(phrase, (counts.get(phrase) ?? 0) + 1);
  }

  const ranked = [...counts.entries()]
    .filter(([p]) => !p.toLowerCase().includes(topic.toLowerCase()))
    .sort((a, b) => b[1] - a[1])
    .map(([p]) => p)
    .slice(0, 4);

  if (ranked.length) {
    try {
      const entry = JSON.stringify({ topic, followups: ranked, at: Date.now(), outcome: "pending" });
      await redis.lpush(suggestionsKey(uid), entry);
      await redis.ltrim(suggestionsKey(uid), 0, 49);
      // Track sources we've shown so future discovery can favor novel ones.
      for (const s of sources) if (s.source) await redis.sadd(seenSourcesKey(uid), s.source.toLowerCase());
      for (const s of sources) {
        try {
          const host = new URL(s.url).hostname.replace(/^www\./, "");
          if (host) await redis.sadd(seenAuthorsKey(uid), host);
        } catch {
          /* skip bad url */
        }
      }
    } catch {
      /* best effort */
    }
  }
  return ranked;
}

/** Record the user's yes/no on a proposed follow-up (the explicit reward signal). */
export async function recordSuggestionOutcome(
  redis: Redis,
  uid: string,
  followup: string,
  accepted: boolean
): Promise<void> {
  await getHermes(redis as unknown as RedisLike).recordOutcome(uid, followup, accepted);
}

/**
 * Curator step: Hermes self-grades recent outcomes and writes a new versioned
 * discovery strategy (+ optional learned skill + reflection) to the Vault.
 * Runs after a session / on schedule via the "curate" job.
 */
export async function runCurate(redis: Redis, uid: string): Promise<DiscoveryStrategy> {
  return getHermes(redis as unknown as RedisLike).reflectAndEvolve(uid);
}

File: worker_py/loops/l2.py
"""L2 — Self-evolving discovery + self-grading. Port of ``worker/loops/l2.ts``.

Candidate follow-ups are proposed here from a session; the yes/no reward and
versioned strategy/skill evolution are handled by the Hermes module.
"""

import json
import re
from typing import Any, Dict, List
from urllib.parse import urlparse

from redis.asyncio import Redis

from ..hermes import get_hermes
from ..util import now_ms

STOP = {
    "the", "and", "for", "with", "that", "this", "from", "what", "your", "you", "are", "how",
    "why", "new", "now", "best", "top", "vs", "use", "using", "into", "about", "over", "last",
    "days", "people", "say", "said", "want", "users", "guide", "tutorial",
}

_PHRASE_RE = re.compile(r"\b([A-Z][a-zA-Z0-9.+#]+(?:\s+[A-Z][a-zA-Z0-9.+#]+){0,2})\b")


def _suggestions_key(uid: str) -> str:
    return f"suggestions:{uid}"


def _seen_sources_key(uid: str) -> str:
    return f"seen_sources:{uid}"


def _seen_authors_key(uid: str) -> str:
    return f"seen_authors:{uid}"


async def propose_followups(
    redis: Redis,
    uid: str,
    topic: str,
    sources: List[Dict[str, Any]],
    synthesis: str,
) -> List[str]:
    """Lightweight follow-up extraction from source titles + synthesis (no LLM)."""
    text = " ".join([synthesis, *[f"{s.get('title', '')} {s.get('snippet') or ''}" for s in sources]])
    topic_words = set(re.split(r"\s+", topic.lower()))

    counts: Dict[str, int] = {}
    for m in _PHRASE_RE.finditer(text):
        phrase = m.group(1).strip()
        lower = phrase.lower()
        if len(phrase) < 3:
            continue
        if lower in topic_words:
            continue
        if all(w in STOP for w in re.split(r"\s+", lower)):
            continue
        counts[phrase] = counts.get(phrase, 0) + 1

    entries = [(p, c) for p, c in counts.items() if topic.lower() not in p.lower()]
    entries.sort(key=lambda x: x[1], reverse=True)  # stable: ties keep insertion order
    ranked = [p for p, _ in entries][:4]

    if ranked:
        try:
            entry = json.dumps({"topic": topic, "followups": ranked, "at": now_ms(), "outcome": "pending"})
            await redis.lpush(_suggestions_key(uid), entry)
            await redis.ltrim(_suggestions_key(uid), 0, 49)
            # Track sources we've shown so future discovery can favor novel ones.
            for s in sources:
                if s.get("source"):
                    await redis.sadd(_seen_sources_key(uid), s["source"].lower())
            for s in sources:
                try:
                    host = (urlparse(s.get("url") or "").hostname or "")
                    host = re.sub(r"^www\.", "", host)
                    if host:
                        await redis.sadd(_seen_authors_key(uid), host)
                except Exception:  # noqa: BLE001
                    continue
        except Exception:  # noqa: BLE001
            pass
    return ranked


async def record_suggestion_outcome(redis: Redis, uid: str, followup: str, accepted: bool) -> None:
    """Record the user's yes/no on a proposed follow-up (explicit reward signal)."""
    await get_hermes(redis).record_outcome(uid, followup, accepted)


async def run_curate(redis: Redis, uid: str) -> Dict[str, Any]:
    """Curator step: Hermes self-grades recent outcomes and writes a new versioned
    discovery strategy (+ optional learned skill + reflection) to the Vault."""
    return await get_hermes(redis).reflect_and_evolve(uid)
